File: C/第3章 链表/3.5 逆序输出链表/inverse.c
#include <stdio.h>
#include "linked_list.h"
#include "stack.h"

void inverse(Node *head) {
    if(head) {
        inverse(head->next);
        printf("%d ", head->data);
    }
}

void inverseNonRecursive(List *list) {
    Stack *s = initStack(getListLength(list));

    Node *node = list->head->next;
    while(node) {
        push(s, node->data);
        node = node->next;
    }

    while(!isEmptyStack(s)) {
        printf("%d ", pop(s));
    }
    printf("\n");

    deleteStack(s);
}

int main() {
    List *list = initList();

    int n = 6;
    Node *nodes[n];
    for(int i = 0; i < n; i++) {
        nodes[i] = (Node *)malloc(sizeof(Node));
        nodes[i]->data = i;
        nodes[i]->next = NULL;
        insert(list, i, nodes[i]);
    }

    inverse(list->head->next);
    printf("\n");

    inverseNonRecursive(list);

    deleteList(list);
    return 0;
}

File: code/Python/Chapter3/3.6 反转链表/inverse.py
import node
import linked_list
import stack

def inverse(node):
    if node:
        inverse(node.next)
        print(node.data, end=' ')

def main():
    lst = linked_list.LinkedList()

    n = 6
    for i in range(n):
        lst.insert(i, node.Node(i))
    
    inverse(lst.get_head())
    print()

if __name__ == "__main__":
    main()